sklearn/preprocessing/_encoders.py

```diff
@@ -0,0 +1,850 @@
+
+    @categorical_features.setter
+    def categorical_features(self, value):
+        warnings.warn("Deprecated", DeprecationWarning)
+        self._deprecated_categorical_features = value
+
+    # Deprecated attributes
+
+    @property
+    def active_features_(self):
+        check_is_fitted(self, 'categories_')
+        warnings.warn("Deprecated", DeprecationWarning)
+        return self._active_features_
+
+    @property
+    def feature_indices_(self):
+        check_is_fitted(self, 'categories_')
+        warnings.warn("Deprecated", DeprecationWarning)
+        return self._feature_indices_
+
+    @property
+    def n_values_(self):
+        check_is_fitted(self, 'categories_')
+        warnings.warn("Deprecated", DeprecationWarning)
+        return self._n_values_
+
+    def _handle_deprecations(self, X):
+
+        if self.categories != 'auto':
+            self._legacy_mode = False
+
+        elif self._deprecated_n_values != 'auto':
+            warnings.warn("Deprecated", DeprecationWarning)
+
+            # we internally translate this to the correct categories
+            # and don't use legacy mode
+            X = check_array(X, dtype=np.int)
+
+            if isinstance(self._deprecated_n_values, numbers.Integral):
+                n_features = X.shape[1]
+                self.categories = [
+                    list(range(self._deprecated_n_values))
+                    for _ in range(n_features)]
+                n_values = np.empty(n_features, dtype=np.int)
+                n_values.fill(self._deprecated_n_values)
+            else:
+                try:
+                    n_values = np.asarray(self._deprecated_n_values, dtype=int)
+                    self.categories = [list(range(i))
+                                       for i in self._deprecated_n_values]
+                except (ValueError, TypeError):
+                    raise TypeError(
+                        "Wrong type for parameter `n_values`. Expected 'auto',"
+                        " int or array of ints, got %r".format(type(X)))
+
+            self._n_values_ = n_values
+            n_values = np.hstack([[0], n_values])
+            indices = np.cumsum(n_values)
+            self._feature_indices_ = indices
+
+            self._legacy_mode = False
+
+        else:  # n_values = 'auto'
+            if self.handle_unknown == 'ignore':
+                # no change in behaviour, no need to raise deprecation warning
+                self._legacy_mode = False
+            else:
+
+                # check if we have integer or categorical input
+                try:
+                    X = check_array(X, dtype=np.int)
+                except ValueError:
+                    self._legacy_mode = False
+                else:
+                    warnings.warn(WARNING_MSG, DeprecationWarning)
+                    self._legacy_mode = True
+
+        if self._deprecated_categorical_features != 'all':
+            self._legacy_mode = True
```

Comment:
This isn't safe to do if `categories` has been set

Reply:
Yes, I know, but I don't want to implement this ability for the new behaviour. 
And, either you were using OneHotEncoder already, and then `categories` was not set and this is OK, or either you did already update your usage (eg setting `categories` instead of n_values) but then you can directly update for this deprecated keyword as well, or either you are new to this class and then you shouldn't use it.

What I can do is detect that `categories` is set by the *user* (and not internally set), and in that case just raise a plain error here instead of a warning.

Reply:
Absolutely. Error if categorical_features is set and not legacy_mode